hazelcast/src/main/java/com/hazelcast/map/impl/record/DataRecordFactory.java

```diff
@@ -65,5 +72,35 @@ public void setValue(Record<Data> record, Object value) {
             v = serializationService.toData(value, partitionStrategy);
         }
         record.setValue(v);
+        updateValueMetadata(record, v);
+    }
+
+    private void updateValueMetadata(Record<Data> record, Data dataValue) {
+        try {
+            Object valueMetadata = metadataInitializer.createFromData(dataValue);
+            Metadata existing = record.getMetadata();
+            if (existing != null) {
+                existing.set(false, valueMetadata);
+            }
+        } catch (Exception e) {
+            rethrow(e);
+        }
+    }
+
+    private Metadata initializeMetadata(Data key, Data value) {
+
+        try {
+            Object keyMetadata = metadataInitializer.createFromData(key);
+            Object valueMetadata = metadataInitializer.createFromData(value);
+            if (keyMetadata != null || valueMetadata != null) {
```

Comment:
Maybe this is thinking too far ahead but is it possible that we have a metadata initializer which for some keys and values will return `null` but might return `non-null` for other values?

Reply:
Yes we can and this code already allows it. In the current state, an entry may have Metadata or not. If it has Metadata, it does not have to have both key and value metadata. It is the getter which tolerates the absence of one (or both). At least, that was my intention. Am I missing something?

Reply:
Nevermind, it's just nitpicking. Let's say we have some metadataInitializer which for the initial combination of key and value returns no metadata. We don't create the `Metadata` instance and proceed by setting the metadata field to `null`.
Later on, the entry might be updated to some other value. Now, the metadata initializer might actually return a non-null valueMetadata. But the metadata field is null and we get a NPE.

Reply:
I see your point. I updated record factoryies' `updateValueMetadata` methods to create metadata if it does not exist. See `DataRecordFactoryTest#testMetadataIsCreatedLaterWhenValueIsUpdatedAndItHasMetadataFirstTime`